fuel/datasets/hdf5.py

```diff
@@ -316,11 +322,21 @@ def _out_of_memory_get_data(self, state=None, request=None):
             if isinstance(request, slice):
                 request = slice(request.start + subset.start,
                                 request.stop + subset.start, request.step)
+                data = state[source_name][request]
             elif isinstance(request, list):
                 request = [index + subset.start for index in request]
+                try:
+                    data = state[source_name][request]
+                except TypeError:
```

Comment:
I feel we should do one of two things here: Set an internal flag `sort_needed` on the first failure, and don't try again afterwards, or allow a keyword argument to be set that does the same thing.

The problem here is that I'm very worried about the speed of the current solution. Try and except-statements can be quite slow, and are usually discouraged if you know that they are triggered often (as could be the case here). I'm also not sure at what point h5py complains, but if that is halfway through the reading, it could be very, very slow.